# Temporal recording of mammalian development and precancer
## Islam et al. 
![Graphical_abstract_v3](https://github.com/Ken-Lau-Lab/NSC-seq/assets/54040073/a96bc4da-b0fc-4c8f-9550-752ded3f032f)

# NSC-seq
## Native sgRNA Capture and sequencing (NSC-seq).

![NSCseq_summary_plot](https://github.com/Ken-Lau-Lab/NSC-seq/assets/54040073/c6fa7ddb-8eff-4592-9430-dfade3b0b5a1)

#### This repository is under development. We will update the data analysis pipeline and code.  

# Here we optimized a multi-purpose single-cell CRISPR platform for: 
## 1. Single-cell lineage tracking 
### We leveraged a self-mutating CRISPR system (e.g. homing CRISPR barcoded MARC1 mouse) to perform lineage tracking at single-cell resolution from embryo to adult stages. 

## 2. Single-cell temporal recording 
### Here we developed a new metric called 'mutation density' by utilizing the accumulative mutation pattern of homing CRISPR to record in vivo temporal events (like cellular turnover). 

## 3. Single-cell CRISPR screen
### Using conventional (KO) vector libraries, we performed single-cell CRISPR screen (Perturn-seq). 
#### Note that the single-cell CRISPR screen application is submitted as a companion manuscript.  

## 4. Clonal analysis of human and murine precancer.  
### Here we perform single-cell multi-omic profiling of human precancers as part of the Human Tumor Atlas Network (HTAN), including 116 scRNA-seq datasets and clonal analysis of 418 human polyps to reveal the clonal composition of human sporadic colorectal precancers. In addition, we validated the polyclonal composition of human precancer using developmentally barcoded mouse mode and NSC-seq platform.  

# Our multimodal framework augments existing single-cell analyses and lays the foundation for in vivo multimodal recording, enabling the tracking of lineage and temporal events during development and tumorigenesis.
